Add support for targeting iOS.

freakboy3742 opened the pull request:
Adds the modifications required to support targeting iOS.

Assumes meson is running in a cross-platform virtual environment, where `sys.platform == "ios"`, and other platform identification tags are "mocked" versions of an iOS environment (matching what cibuildwheel's iOS support does).

Functionally, it's very similar to macOS build support - it writes a cross-build configuration file, based on the contents of sysconfigdata.

The bump in the version of Packaging is required so that iOS platform tag handling is available.

dnicolodi reviewed (changes requested):
Thanks for working on this. I don't know anything about building for iOS: the use of environment variables to configure the build goes in the opposite direction of where I would like the ecosystem to go. I can be convinced that this is the only sane way, but the handling of the `CFLAGS` and `BLDSHARED` environment variables is IMO wrong.

rgommers commented on pyproject.toml:
```diff
@@ -8,7 +8,7 @@ backend-path = ['.']
   'meson >= 0.64.0; python_version < "3.12"',
   'meson >= 1.2.3; python_version >= "3.12"',
-  'packaging >= 19.0',
+  'packaging >= 24.2',
```
> Does iOS have an environment marker assigned?

If not, now would be a great time to add it, since PEP 780 is the first time in a long time that we're trying to add new environment markers.

rgommers replied:
Pip upgraded its vendored `packaging` to 25.0 already, so we _might_ we able to get away with this bump, but it's hard to be sure so we'd rather not have to do this bump to a 6 months old release.

dnicolodi replied:
`sys_platform` should work, I think:
```toml
requires = [
  '...',
  'packaging >= 19.0; sys_platform != "ios"',
  'packaging >= 24.2; sys_platform = "ios"',
  '...',
]
```

freakboy3742 replied:
iOS platform markers were defined as part of PEP 730; they've been incorporated into pip and packaging (hence the version bump).

dnicolodi replied:
What I meant is to use some env markers in the dependencies specification to require the newer `packaging` only when building for iOS, like in the snippet I proposed above. I think that dependencies specification should work also wit Python/pip/packaging that do not know yet about iOS.

dnicolodi commented on tests/test_project.py:
```diff
@@ -373,3 +376,40 @@ def test_archflags_envvar_parsing_invalid(package_purelib_and_platlib, monkeypat
         os.environ.pop('_PYTHON_HOST_PLATFORM', None)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 13), reason="Requires Python 3.13 or higher")
```
Nit: I think we are standardizing to test skip reasons starting with a lowercase character (they look a bit nicer in pytest output).

dnicolodi commented on tests/test_tags.py:
```diff
@@ -74,6 +75,23 @@ def test_python_host_platform(monkeypatch):
     assert mesonpy._tags.get_platform_tag().endswith('x86_64')
 
 
+@pytest.mark.skipif(sys.version_info < (3, 13), reason="Requires Python 3.13 or higher")
```
Here too.

freakboy3742 commented on mesonpy/__init__.py:
```diff
@@ -717,6 +718,62 @@ def __init__(
+            arch = platform.machine()
+            family = 'aarch64' if arch == 'arm64' else arch
+            subsystem = 'ios-simulator' if ios_ver.is_simulator else 'ios'
+            min_ios_version = os.getenv('IPHONEOS_DEPLOYMENT_TARGET', ios_ver.release)
```
Yes - it's a [standard iOS environment variable](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW156), the analog of `MACOSX_DEPLOYMENT_TARGET` for macOS.

rgommers replied:
Then we should treat it the same way indeed. Can you add it to the docs in the same way as `MACOSX_DEPLOYMENT_TARGET`?

https://github.com/mesonbuild/meson-python/blob/e094949f466ae69043fcec3ec68b41ace900ed97/docs/reference/environment-variables.rst#L70

rgommers replied:
@freakboy3742 a gentle ping on this one in case you missed it - I think it's last comment to address here.

freakboy3742 replied:
My apologies - I did miss it; I assumed that because the code no longer referenced the environment variable, it was no longer relevant. However, looking closer, I see that MACOSX_DEPLOYMENT_VERSION isn't used in the *code* either, but it's obviously significant there as well.

I've pushed a documentation update (with a matching grammar tweak to the MACOSX_DEPLOYMENT_VERSION docs).

freakboy3742 replied:
@dnicolodi Apologies for the confusion - there's a few crossed wires here. 

I originally didn't have documentation for `IPHONEOS_DEPLOYMENT_TARGET` because it wasn't being used *explicitly* by meson-python; however, @rgommers pointed out that it *is* used - just indirectly, as part of the wheel tag. My observation was that this is the same situation as `MACOSX_DEPLOYMENT_TARGET`, so I added docs for `IPHONEOS_DEPLOYMENT_TARGET`, and tweaked the corresponding docs for macOS to highlight the `-mmacosx-version-min` limitation.

freakboy3742 commented on mesonpy/__init__.py:
```diff
@@ -717,6 +718,62 @@ def __init__(
+            for flag in shlex.split(sysconfig.get_config_var('CFLAGS')):
+                # Ensure the iOS minimum version reflects the current build environment
+                if flag.startswith('-mios-version-min='):
+                    cflags.append(f"-mios-version-min={min_ios_version}")
```
This isn't guessing or contradicting anything - it's expressing specified preferences. 

CPython can be (and is) built with a minimum iOS version (13.0 by default), and this is reflected in the compiler flags that are contained in sysconfig data (in this case, CFLAGS).

However,  an individual project can then be compiled with a *different* minimum iOS version. For example, if you want to build NumPy with full accelerated math, you need a min iOS version of 16.4. This obviously results in a binary that can only be used on iOS 16.4 - but that can be (and is) reflected in the iOS wheel tag for the resulting binaries.

So - this code is ensuring that the minimum iOS version used for compilation is the one specified as a part of this Meson build, not the one specified at the *CPython* level.

dnicolodi replied:
I think I understand.

In your picture, the `$CFLAGS` env var is set to what Python uses for its compilation taken from `sysconfig` data and Python has a `-mios-version-min=` flag in there, while the user intent for a minimum iOS version is reflected in the `$IPHONEOS_DEPLOYMENT_TARGET` env var. Therefore, you make sure that the former is adjusted to the latter.

In my picture, both `$CFLAGS` and `$IPHONEOS_DEPLOYMENT_TARGET` come from the user. The guessing  I was referring to is relative to deciding which environment variable should take precedence in defining the min iOS version: if you assume that the user is responsible for the value of both environment variables, you can only guess which one is the one the user intended to be correct and the best thing to do is to error out to indicate that the user specified an inconsistent configuration.

With this sorted, I have a few questions regarding the handling of `$IPHONEOS_DEPLOYMENT_TARGET` and `-mios-version-min=`.

Why does CPython adds `-mios-version-min=` to the compilation flags required to link with it? IIUC how this is works at the OS level, it does not seem necessary. What does the compiler do when the `$IPHONEOS_DEPLOYMENT_TARGET` env var is set and `-mios-version-min=` is specified? What happens when `-mios-version-min=` appears multiple times on the command line? Having `-mios-version-min=` in the compilation flags in sysconfig (and thus exposed by `python-config`, I guess) does not only seem unnecessary, but a potential source of issues.

When compiling for macOS with do not fiddle with `-mmacosx-version-min=` compilation flag and rely entirely on the `$MACOSX_DEPLOYMENT_TARGET` env var, also for deriving the platform tag. This is sub optimal as the user has many ways to pass `-mmacosx-version-min=` to the build and that would not be reflected in the wheel platform tag. I'm not sure about the best way forward here. Maybe continuing to rely on the env vars alone here is the best think, at least for continuity. However, if as expected the command line flag takes precedence over the env var, Meson should not include the `-mios-version-min=` when pickling up the compilation flags from Python.

freakboy3742 replied:
These are all good and reasonable questions; at this point, we're hitting up against "things Apple doesn't document well because Why Aren't You Just Using Xcode", so I'll need to do some experimentation to confirm what I *think* is true is actually the behaviour in practice.

tl;dr is that "just use environment variables" might be a viable approach, but I've got at least one case that doesn't appear to be behaving quite right in my testing - I need to confirm that's not operator error.

freakboy3742 replied:
After some experimentation (and a lot of hair-pulling), I've been able to get to the bottom of things here:

1. `clang` has *three* ways to control the minimum SDK: 
    - the `IPHONEOS_DEPLOYMENT_TARGET` environment variable
    - the `-mios-version-min` compiler argument, and
    - a version number encoded into the `-target` compiler triple (e.g., `-target arm64-apple-ios13.0`)
2. If `-mios-version-min` will override the environment variable without any warning
3. If multiple `-mios-version-min` arguments are specified, the last one on the command line takes priority
4. If `-mios-version-min` *and* `-target` are specified, `-target` takes priority; and if the version number doesn't match, a warning is raised.
5. If a "version unspecified" target is provided (e.g., `-target arm64-apple-ios`), this is interpreted as "no minimum version". A `-mios-version-min` will override this - but `IPHONEOS_DEPLOYMENT_TARGET` will *not* - an environment variable is *ignored* in this situation.
6. The option doesn't appear go be needed on link; that looks like I was being overzealous in flag usage.

This actually revealed a bug in CPython's compiler shims (see python/cpython#133183) - the compiler shims were effectively overriding the `IPHONEOS_DEPLOYMENT_TARGET` provided in the environment. This is essentially why I included the `-mios-version-min` argument in the cross build file - without the explicit `-mios-version-min` argument, `numpy` would fail to build when it hit a language feature that wasn't available before iOS 7, even if `IPHONEOS_DEPLOYMENT_TARGET` was set in the environment.

---

So - the tl;dr is that with a fix in CPython's shims to address the environment variable issue, and promoting the Python library linking argument upstream to meson, the iOS cross-file can be reduced to the pure compiler definitions (just the binary names), and the platform identifier section that will be common to all iOS builds.

I'll push an update as soon as I've got all the pieces sorted out.

dnicolodi replied:
Thanks for investigating this. I'm not sure I understand all the details:

> 5\. If a "version unspecified" target is provided (e.g., `-target arm64-apple-ios`), this is interpreted as "no minimum version". A `-mios-version-min` will override this - but `IPHONEOS_DEPLOYMENT_TARGET` will _not_ - an environment variable is _ignored_ in this situation.

Are you saying that when the `-taget arm64-apple-ios` build flag is specified and no `-mios-version-min` is specified, the `IPHONEOS_DEPLOYMENT_TARGET` env var is ignored? Given points 2 and 4 above, that would mean that the environment variable is always ignored. I think I'm misunderstanding something.

freakboy3742 replied:
Yes, that is the consequence.

The other piece of the puzzle is that the `-target` option isn't the only way to drive `clang` to build for iOS. The Apple-endorsed way to invoke `clang` is `xcrun --sdk iphoneos clang`; Apple then provides 2 ways to drive clang to produce an iOS binary for a specific architecture:
* `xcrun --sdk iphoneos clang -target arm64-apple-ios`. 
* `xcrun --sdk iphoneos clang -arch arm64`

In the second usage, the `IPHONEOS_DEPLOYMENT_TARGET` environment variable *would* be honoured; to target the simulator, you'd need to pass in `--sdk iphonesimulator`. 

This isn't well documented, but it *appears* that `-target` is Apple's preferred option (to the extent they express preferences). Aside from being more explicit; when compiling for visionOS, there is no `-mvisionos-version-min` (or equivalent) argument to clang; the version specifier *must* be provided with the `-target` triple, or with `XROS_DEPLOYMENT_TARGET` . There's only 1 architecture for visionOS, so the `-target` is *strictly* unnecessary... I think? It's difficult to prove categorically, and Apple doesn't document this stuff well - again, "Why Aren't You Using Xcode?")

tl;dr - Apple has far too many options for the same thing :-)

rgommers replied:
This is very useful content for some future debugging I'm sure, so let's leave this comment thread unresolved.

freakboy3742 commented on mesonpy/__init__.py:
```diff
@@ -717,6 +718,62 @@ def __init__(
+
+            ldflags = []
+            # BLDSHARED includes the linker as the first argument; we just need the flags.
+            for flag in shlex.split(sysconfig.get_config_var('BLDSHARED'))[1:]:
```
`BLDSHARED` is a [standard setting in `sysconfig` data](https://github.com/python/cpython/blob/main/configure.ac#L1389). It's the same as `LDSHARED` on most platforms, but [has some important differences on macOS platforms](https://github.com/python/cpython/blob/main/configure.ac#L1527).

freakboy3742 replied:
That said - `sysconfig.get_config_var("prefix")` is a better source for the one "variable" detail we need here, so constructing `c_link_args` et al from "first principles" is probably a better approach.

mhsmith replied:
I see this has all changed now, but for future reference: my understanding is that BLDSHARED and BLDLIBRARY are only supposed to be used for building Python itself. Third-party extension modules should use the corresponding variables without the L prefix. If this distinction has been blurred in macOS or iOS, then maybe that should be revisited on the CPython site.

freakboy3742 replied:
I don't believe the distinction has been blurred. As noted, it's moot now; but I think I was overthinking the problem in what I was trying to exclude in the flags captured in the cross file.

freakboy3742 commented on mesonpy/__init__.py:
```diff
@@ -717,6 +718,62 @@ def __init__(
+                longdouble_format = 'IEEE_DOUBLE_LE'
+            ''')
+            self._meson_cross_file.write_text(cross_file_data, encoding='utf-8')
+            self._meson_args['setup'].extend(('--cross-file', os.fspath(self._meson_cross_file)))
```
I'll freely admit I'm new to Meson - my experience is limited to getting Numpy to compile on iOS - so I'm definitely open to suggestions of a better way of handling iOS builds that are more in alignment with "best practices" for Meson usage.

However, unless I'm misunderstanding what you're suggesting, or I'm missing some detail of how Meson is used in practice, I'm don't agree that expecting end-projects to pass in a cross-build file *is* a better solution for the iOS case. 

iOS (and, for that matter macOS), are fundamentally different environments to Unix environments, because they're able to provide guarantees that Unix environments cannot - binary and build tool compatibility. 

If you've got a binary that will run on a 6 year old iOS device running iOS 13, I can *guarantee* that binary will run, unmodified, on a brand new iPhone 16 running iOS 18. I can *guarantee* the full set of system libraries that are available, and I can *guarantee* the third-party libraries that are available - because they *must* be bundled with the app (as there's no "shared between app" resources that aren't system-provided).

And, on top of that - the compiler binary and the flags passed to that binary will be *identical*, regardless of which device you're running on. This is all controlled by two things - the `IPHONEOS_DEPLOYMENT_TARGET` (also specified as `-mios-min-version`), which guarantees the oldest iOS version where the code will run; and the SDK version that is being used to compile the toolchain. Yes, there are additional flags and options that can be enabled; but they're the sort of compiler flags that are project dependent (e.g, whether the project enables `-Wall`), or flags that enable specific features that are required for the project.

So - unlike a Unix environment, we *are* in a position to provide a "canonical" iOS cross-build file that encodes the basic flags for how to compile *every* Python project for iOS:

* The C compiler is known (e.g, `arm64-apple-ios-simulator-cc`). [This are shims provided as part of a Python iOS environment](https://github.com/python/cpython/tree/main/iOS/Resources/bin)
* `-mios-version-min=<ver>` *must* be used on every compilation. CPython itself is defined with a minimum iOS version, so anything built to use with CPython also needs this flag. The version specified *can* be newer, but it can't be older than the version used by CPython.
* `-F <path to python lib> -framework Python -mios-version-min=<ver>` *must* be specified on every link. Meson will also add `-dynamiclib`.
* The `host_machine` definition will be identical.

With all that said: reflecting on your feedback, I agree that manually filtering and updating `CFLAGS` and `BLDFLAGS` isn't an ideal approach. The values will be reasonably stable and predictable - as they are derived from how CPython itself is compiled - but they can be explicitly derived from "first principles" and other more authoritative sources. I'll update the code to reflect this.

As for whether this is bound the implementation used by cibuildwheel - not really. The environment used by cibuildwheel is composed of (a) a stock CPython compilation for iOS, plus (b) modifications to set up a cross-platform build environment. The *implementation* of the cross-platform environment is cibuildwheel/BeeWare specific - but is also largely irrelevant. There are multiple approaches that can be used to generate a cross-plaform environment, but the only pieces that this patch is depending on are features that exist in *any* virtual environment - properties like `sys.platform`, methods in the `platform` module, and keys in `sysconfig` data. Any virtual environment that behaves like a "normal" Python environment, but exposes iOS-specific values reflecting local conditions, should yield the same result.

eli-schwartz replied:
> If you've got a binary that will run on a 6 year old iOS device running iOS 13, I can _guarantee_ that binary will run, unmodified, on a brand new iPhone 16 running iOS 18. I can _guarantee_ the full set of system libraries that are available, and I can _guarantee_ the third-party libraries that are available - because they _must_ be bundled with the app (as there's no "shared between app" resources that aren't system-provided).

(You could argue that this maps to Linux, inasmuch as the full set of system libraries that are available is libc, libm, libpthread, librt, libresolv... and the system libraries never break ABI, and you should simply bundle anything else you need. Admittedly, you cannot pass a -mminimum-glibc-version.)

> `-F <path to python lib> -framework Python -mios-version-min=<ver>` _must_ be specified on every link. Meson will also add `-dynamiclib`.

If meson itself does not yet pass `-F <path to python lib> -framework Python` then we probably should. Hmm, this is probably relevant:
https://github.com/mesonbuild/meson/blob/2e8a804a009d5ee0c4bfba8bf629ace3a28c6e14/mesonbuild/dependencies/python.py#L441-L447

We don't normally expect people to pass `CFLAGS="-I/usr/include/python3.13"` on linux, we should not expect people to pass the framework flags in CFLAGS on iOS either.

freakboy3742 replied:
> > `-F <path to python lib> -framework Python -mios-version-min=<ver>` _must_ be specified on every link. Meson will also add `-dynamiclib`.
> 
> If meson itself does not yet pass `-F <path to python lib> -framework Python` then we probably should. Hmm, this is probably relevant: https://github.com/mesonbuild/meson/blob/2e8a804a009d5ee0c4bfba8bf629ace3a28c6e14/mesonbuild/dependencies/python.py#L441-L447

I was under the impression that Meson wasn't Python specific - evidently I've missed something here in my analysis. Poking around that file, I think I can see where linking flags can be included - I'll see what I can work out.

> We don't normally expect people to pass `CFLAGS="-I/usr/include/python3.13"` on linux, we should not expect people to pass the framework flags in CFLAGS on iOS either.

That makes sense - it hadn't occurred to me that in all my iOS working, I'd never needed to pass in `-I`. It's obvious in retrospect.

eli-schwartz replied:
> I was under the impression that Meson wasn't Python specific - evidently I've missed something here in my analysis. Poking around that file, I think I can see where linking flags can be included - I'll see what I can work out.

Possibly you mean in the sense that meson can target an arbitrary python -- but it does that here: https://github.com/mesonbuild/meson/blob/master/mesonbuild/scripts/python_info.py

Any information that the dependency lookup needs to know about can be probed like this via the installation info: https://github.com/mesonbuild/meson/blob/2e8a804a009d5ee0c4bfba8bf629ace3a28c6e14/mesonbuild/dependencies/python.py#L391-L396

freakboy3742 replied:
Sorry - I meant in the sense that while it is a build tool that is built using Python, it wasn't explicitly a tool for *building* Python projects - it's as suitable for building a "hello world written in C" project as it would be a Python one.

The bit I was missing was that "Python" is a known dependency that a Meson project can declare, and once that dependency is declared, a bunch of flags (like include paths and link flags) can be inferred and used as part of the build.

eli-schwartz replied:
Aha, well, meson's python support is (from meson's perspective) "just another [optional module](https://mesonbuild.com/Modules.html)" :) and we're always happy to work with people to enable new use cases.

There's a lot of minutiae that goes into building python projects, e.g. meson directly knows how to probe for `EXT_SUFFIX` and use it to name compiled extensions too. We want to collect that information in one place, for several reasons including the fact that there are people building applications which [embed python](https://docs.python.org/3/extending/embedding.html) rather than extending it, and other people who are building extensions that aren't meant to be installed as wheels, but only via a linux package manager, such as https://gitlab.gnome.org/GNOME/gobject-introspection and its `giscanner/` (yes, this includes a python C extension).

In fact, meson can directly lay out a python project including installing all its `*.py` and `*.so` files correctly into site-packages/ too. `meson-python` just handles converting that into a wheel.

> The bit I was missing was that "Python" is a known dependency that a Meson project can declare, [...]

So yeah, details about building python projects which can be unambiguously correct are potentially a candidate to be handled directly in meson, and "how to correctly compile and link against a specific python dependency" is top of the list, here. Although e.g. things like synthesizing a cross file specification or calculating a minimum ios version are likely better handled in meson-python.

dnicolodi replied:
I'll try to clarify my comments. There are a few different aspects to consider.

First, should meson-python synthesize a cross build file, or should the user be responsible for providing one?

The "Meson native way to setup a cross build is to pass a cross build definition file ("cross file" for short, and from the name of the option used to specify it) , namely a specification of the host machine and of the compilers that should be used for the compilation, see https://mesonbuild.com/Cross-compilation.html

This works also for cross compiling Python packages using meson-python and it is how conda-forge builds packages for some Linux platforms. This works something like this:
```
python -m build -Csetup-args=--cross-file=$CROSS_FILE_PATH
```
What meson-python does on macOS when the `$ARCHFLAGS` environment variable is set is to synthesize the cross file and pass that flag to Meson. This was implemented to be compatible with the (ugly) UX provided by setuptools to cross-compile on macOS (there are historical reasons why setuptools works like that which explain why the UX is so ugly, but they are not relevant here). IMO, environment variables are a suboptimal way to configure aspects of a project compilation (and most other aspects of any piece of software).

I was wondering whether it makes sense to rely on environment variables to configure a cross build for iOS, or it is better to require users to write a cross file and pass it at build time like above. If the cross file is always the same (because the cross compilation can happen only on macOS and the cross build tools are setup always in the same way) this only implies that the cross file can be written once and reused.

However, I realize that native compilation for iOS is not really a thing, thus it seems to be sensible to have meson-python synthesize one when it detects that it is targeting a iOS Python, as that is the only possible way to compile for this platform.

Second, what aspects of the compilation should be configured via the generated cross file?

From the code and the comments on the PR, I understand that the proposed approach adds to the cross file compilation flags specific to building a Python extension module. As hinted by @eli-schwartz, I think this is not the best approach: compilation flags required to build a Python extension should be discovered by the Meson's `python` module. Furthermore, it is not a good UX to expect the user to have to setup `$CFLAGS` and `$BLDFLAGS` just right if the flags that should go in there can be derived from introspecting the Python interpreter, and because environment variables are not a good way to configure builds.

Therefore, I hope we can avoid having to inspect `$CFLAGS` and `$BLDFLAGS` in meson-python.

I hope this makes sense so far. This leaves the handling of `$IPHONEOS_DEPLOYMENT_TARGET`, see below for more comments on that.

freakboy3742 replied:
Yes, this makes sense; thanks for the clarifications. I need to do some more poking to get the link flags inserted upstream, and to work out what's going on with min versions; I'll report back when I've got all the details straight.

freakboy3742 commented on mesonpy/__init__.py:
```diff
@@ -717,6 +718,62 @@ def __init__(
+                    cflags.append(f"-mios-version-min={min_ios_version}")
+                # Don't propagate warning settings or preprocessor defines
+                elif not (flag.startswith('-W') or flag.startswith('-D')):
+                    cflags.append(flag)
```
I can't say I've done extensive testing of the interaction between environment CFLAGS and this setting. 

As noted in the earlier comment - Unlike Unix environments, there's very little need for "local environment specific" configuration of compilations, because an iOS development environment is essentially fixed - every iOS device is essentially identical, inside the bounds of minimum and maximum supported iOS versions. 

However, in the case of CFLAGS specifically, the only flag we *must* have is `-mios-version-min`, which *could* be incorporated into the `c`/`cpp`/etc definition.

eli-schwartz replied:
> (I assume environment variables have higher priority as source of configuration in Meson than the content of cross files, but I haven't verified it).

meson considers cross files to be a specific statement of overriding intent, they are ignored if you set the corresponding cross file value.

rgommers replied:
`CFLAGS` modification is gone, so I'll resolve this comment.

freakboy3742 commented on mesonpy/_tags.py:
```diff
@@ -160,10 +160,25 @@ def _get_macosx_platform_tag() -> str:
+    except ValueError:
+        version = tuple(map(int, platform.ios_ver().release.split('.')))[:2]
+
+    multiarch = sys.implementation._multiarch.replace('-', '_')
```
In the case of iOS: yes, it's a guaranteed part of the implementation, containing `arm64-iphonesimulator` or `arm64-iphoneos`, depending on the intended target. As for prior art... it's used [in CPython itself as part of iOS platform identification](https://github.com/python/cpython/blob/main/Lib/platform.py#L898).

dnicolodi replied:
I was hoping to have some confirmation that underscore prefixed attributes in `sys.implementation` is considered public API, despite the Python naming convention, and that relying upon them is ok. Use of these attributes withing CPython itself is not relevant because any change to the presence,  meaning, or format of these attributes would be updated in sync within the CPython codebase.

Is `pckaging` using similar code as the above to derive platform flags?

rgommers replied:
`_multiarch` was removed from [PEP 739](https://peps.python.org/pep-0739/) during the review phase (it's still shows in one example, but that's a mistake - it's not in the spec nor in the JSON Schema for it).

I think it's pretty stable but technically private to CPython.

dnicolodi replied:
`packaging` uses the same field https://github.com/pypa/packaging/blob/fe7dd6d8e8d938acacad205692b6729b0368d505/src/packaging/tags.py#L496-L498 Thus at least we will not be the only ones complaining if this break. This is good enough for me.

rgommers replied:
Okay good - agreed that that should work fine then. A code comment explaining this or referencing this comment thread may be useful for future reference.

rgommers commented on mesonpy/__init__.py:
```diff
@@ -717,6 +717,40 @@ def __init__(
+                endian = 'little'
+
+                [properties]
+                longdouble_format = 'IEEE_DOUBLE_LE'
```
This property is numpy-specific. I don't mind hardcoding that in NumPy (skip the runtime check if we're cross-compiling to iOS), but we shouldn't need to add it here.

freakboy3742 commented:
I've just pushed an update that (I think) addresses the concerns raised about the previous version; in summary:

* All linking additions are now handled upstream; see mesonbuild/meson#14541 for the corresponding addition
* The handling of iOS min API versions are now deferred to environment variables and/or individual project configurations; see python/cpython#133184 for the upstream changes to the compiler shims to ensure this works for all cases
* the handling of `longdouble_format` has been pushed down to numpy (where it originated); see [this commit](https://github.com/numpy/numpy/pull/28759/commits/bf0802b966a30bc7a974e4c279a9c4c9958a2fa1)

This means the cross-build file is now fairly minimal - it requires no environment lookups, and only specifies things that are "core" to iOS - the compiler names, and the platform identification tags.

There's a lot of moving pieces here to be able to test; I've tested it with my numpy PR, but it requires patched versions of  numpy, CPython, and meson, in addition to these changes.

freakboy3742 commented:
The test failures here seem unrelated; the fedora failure is a CI credits issue; the macOS issue looks like it might be related to the recent meson 1.8 release?

freakboy3742 commented:
The two upstream PRs (python/cpython#133184 and mesonbuild/meson#14541) have now been merged.

rgommers commented:
The macOS issue resolved itself, Fedora failure can be ignored. There are two other test failures that are due to this PR @freakboy3742, could you resolve those?

freakboy3742 commented:
> There are two other test failures that are due to this PR @freakboy3742, could you resolve those?

Will do first thing tomorrow. Initial appearances look like some straightforward typing and platform-specific test stuff.

freakboy3742 commented:
> > There are two other test failures that are due to this PR @freakboy3742, could you resolve those?

Done.

1. The typing tests are running on Python 3.9, so the `platform.ios_ver()` method doesn't exist. I've followed the same approach used by `packaging`, which is to `ignore[attr-defined]` on the 2 affected lines.
2. The Windows issue is caused because `sys.implemenation._multiarch` doesn't exist on Windows. It *exists* on macOS... but it doesn't have an iOS-meaningful value, but the inconsistency is at least consistent between `packaging` and `meson-python` so it didn't fail the test. I've added `sys.implemenation._multiarch` to the monkey-patching list, to give a more reliable test of iOS behavior.

freakboy3742 commented:
Grumble... looks like another Windows test failure. Looks like we just skip that test on Windows, then.

rgommers commented:
If you want to submit a separate PR for a typo fix that I can merge straight away, then GHA CI jobs won't need manual approval anymore. There's one on line 763 of `__init__.py` ("metdata").

freakboy3742 commented:
> If you want to submit a separate PR for a typo fix that I can merge straight away, then GHA CI jobs won't need manual approval anymore. There's one on line 763 of `__init__.py` ("metdata").

Done - see #748.

dnicolodi commented on docs/reference/environment-variables.rst:
```diff
@@ -61,6 +61,21 @@ Environment variables used by meson-python
+   Another way of specifying the target iOS platform is to use the
+   ``-mios-version-min`` compile and link flags.  However, it is not possible
+   for ``meson-python`` to detect this, and it will not set the Python wheel
+   platform tag accordingly.
```
I would either remove this paragraph (as it does not add any information useful to meson-python users) or rephrase it: starting with stating that it exist another way to do something, and concluding noting that it does not work is not good style for documentation.

freakboy3742 replied:
This paragraph is a modified version of the one that already exists for `MACOSX_DEPLOYMENT_TARGET`; however, I agree it's worded awkwardly given the context. I'll clean both up.

dnicolodi reviewed (changes requested):
Can you please squash together the commits that make sense to be squashed, and provide a commit message in the style used in this repository?

I've lost track of things: are the patches needed for iOS support merged into Meson and CPython already?

dnicolodi commented:
> I can't find any contribution guide to confirm what the required style is

Indeed, we don't have such a guide. We use the NumPy commit message style https://numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

freakboy3742 commented:
> > the Meson patches have all landed, and are waiting on the 1.9 release.
> 
> Then I think a note should be added here
> ...
> stating that meson 1.9.0 or later is required for iOS support.

Done. 

For complete transparency, 1.9.0 hasn't been released yet, but the required code is all in the main branch, so as soon as a new feature release is made, it should be included.

dnicolodi commented on docs/reference/meson-compatibility.rst:
```diff
@@ -52,6 +52,10 @@ versions.
 
+.. option:: 1.9.0
+
+   Meson 1.9.0 or later is required to support building for ``iOS``.
```
```suggestion
   Meson 1.9.0 or later is required to support building for iOS.
```
iOS is a product name thus I don't think it should be rendered in fixed-width font like code.

dnicolodi commented on docs/reference/environment-variables.rst:
```diff
@@ -84,10 +99,11 @@ Environment variables used by meson-python
-   ``-mmacosx-version-min`` compile and link flags.  However, it is not
-   possible for ``meson-python`` to detect this, and it will not set the
-   Python wheel platform tag accordingly.
+   Note that ``MACOSX_DEPLOYMENT_TARGET`` is the *only* permitted mechanism for
```
```suggestion
   Note that ``MACOSX_DEPLOYMENT_TARGET`` is the only supported mechanism for
```

dnicolodi commented on docs/reference/environment-variables.rst:
```diff
@@ -61,6 +61,22 @@ Environment variables used by meson-python
+   This variable must be set to a major/minor version (e.g., ``13.0`` or
+   ``15.4``).
+
+   Note that ``IPHONEOS_DEPLOYMENT_TARGET`` is the *only* permitted mechanism
```
```suggestion
   Note that ``IPHONEOS_DEPLOYMENT_TARGET`` is the only supported mechanism
```

dnicolodi reviewed (changes requested):
Thanks. Just a couple of minor things to fix in the doc.

dnicolodi commented:
Thanks. I tweaked up a couple of nits myself. Merging.

henryiii commented:
cibuildwheel has shipped iOS support in 3.0 this morning, FYI. Any plans for a release here?

eli-schwartz commented:
The meson bits haven't yet shipped in a meson release, FWIW.
